File: tools/qualification/live_workbench_paint.mjs
import { spawn } from 'node:child_process';
import { createHash } from 'node:crypto';
import { readFile, mkdir, writeFile } from 'node:fs/promises';
import { createServer } from 'node:net';
import { join, resolve } from 'node:path';
import { chromium } from 'playwright-core';

const [binaryArg, fixtureArg, targetArg, outputArg, sampleArg = '20'] = process.argv.slice(2);
if (!binaryArg || !fixtureArg || !targetArg || !outputArg) {
	throw new Error(
		'usage: node live_workbench_paint.mjs BINARY FIXTURE MINIMUM_NODES OUTPUT_DIR [SAMPLES]'
	);
}
const binary = resolve(binaryArg);
const fixture = resolve(fixtureArg);
const outputDir = resolve(outputArg);
const minimumNodes = Number(targetArg);
const samples = Number(sampleArg);
if (
	!Number.isSafeInteger(minimumNodes) ||
	minimumNodes < 1 ||
	!Number.isSafeInteger(samples) ||
	samples < 1
) {
	throw new Error('minimum nodes and samples must be positive safe integers');
}

const hashFile = async (file) =>
	createHash('sha256')
		.update(await readFile(file))
		.digest('hex');
const pause = (ms) => new Promise((done) => setTimeout(done, ms));
const percentile = (values, quantile) => {
	const sorted = [...values].sort((a, b) => a - b);
	return sorted[Math.min(sorted.length - 1, Math.ceil(sorted.length * quantile) - 1)];
};

async function availablePort() {
	const server = createServer();
	await new Promise((done, reject) => {
		server.once('error', reject);
		server.listen(0, '127.0.0.1', done);
	});
	const port = server.address().port;
	await new Promise((done, reject) => server.close((error) => (error ? reject(error) : done())));
	return port;
}

async function request(baseUrl, route, body) {
	const response = await fetch(`${baseUrl}${route}`, {
		method: body === undefined ? 'GET' : 'POST',
		headers: body === undefined ? {} : { 'content-type': 'application/json' },
		body: body === undefined ? undefined : JSON.stringify(body),
		signal: AbortSignal.timeout(180_000)
	});
	const text = await response.text();
	if (!response.ok) throw new Error(`${route} returned ${response.status}: ${text.slice(0, 500)}`);
	return JSON.parse(text);
}

async function waitForHealth(baseUrl, child) {
	const deadline = Date.now() + 60_000;
	while (Date.now() < deadline) {
		if (child.exitCode !== null) throw new Error(`headless app exited with ${child.exitCode}`);
		try {
			const health = await request(baseUrl, '/api/ui/health');
			if (health.backend_ready && health.engine_read_model_ready) return;
		} catch {
			// The listener may not have started yet.
		}
		await pause(100);
	}
	throw new Error('headless app did not become healthy');
}

function chooseDuplicateSources(snapshot, minimumInsertedNodes = 600) {
	const byId = new Map(snapshot.nodes.map((node) => [node.node_id, node]));
	const parentByChild = new Map();
	for (const parent of snapshot.nodes) {
		for (const child of parent.children ?? []) parentByChild.set(child, parent.node_id);
	}
	const roots = snapshot.nodes.filter(
		(node) => typeof node.decl_id === 'string' && node.decl_id.startsWith('scale_constant_')
	);
	const nodes = [];
	let insertedNodes = 0;
	for (const root of roots) {
		const parent = parentByChild.get(root.node_id);
		if (!parent) throw new Error(`authored Constant ${root.decl_id} has no parent`);
		const pending = [root.node_id];
		let subtreeNodes = 0;
		while (pending.length > 0) {
			const node = byId.get(pending.pop());
			if (!node) throw new Error('snapshot has a missing subtree child');
			subtreeNodes += 1;
			pending.push(...(node.children ?? []));
		}
		nodes.push({ source: root.node_id, new_parent: parent });
		insertedNodes += subtreeNodes;
		if (insertedNodes >= minimumInsertedNodes) break;
	}
	if (insertedNodes < minimumInsertedNodes) {
		throw new Error(`only ${insertedNodes} duplicable nodes; need ${minimumInsertedNodes}`);
	}
	return { nodes, insertedNodes };
}

async function run() {
	await mkdir(outputDir, { recursive: true });
	const appData = join(outputDir, 'appdata');
	await mkdir(appData, { recursive: true });
	const port = await availablePort();
	const baseUrl = `http://127.0.0.1:${port}`;
	const child = spawn(binary, ['--headless', '--no-remote'], {
		cwd: resolve('.'),
		env: { ...process.env, APPDATA: appData, GC_UI_BIND: `127.0.0.1:${port}` },
		stdio: ['ignore', 'pipe', 'pipe'],
		windowsHide: true
	});
	let serverLog = '';
	child.stdout.on('data', (chunk) => {
		serverLog += chunk.toString();
	});
	child.stderr.on('data', (chunk) => {
		serverLog += chunk.toString();
	});
	const report = {
		contract: 'chataigne-live-workbench-paint-v1',
		status: 'FAIL',
		minimum_nodes: minimumNodes,
		samples,
		binary_sha256: await hashFile(binary),
		fixture_sha256: await hashFile(fixture),
		browser_version: null,
		base_nodes: null,
		duplicated_roots_per_sample: null,
		inserted_nodes_per_sample: null,
		latencies_ms: [],
		http_ack_ms: [],
		mutation_ms: [],
		outliner_items_per_sample: [],
		outliner_projected_rows_per_sample: [],
		p50_ms: null,
		p95_ms: null,
		p99_ms: null,
		max_ms: null,
		p95_target_ms: minimumNodes === 10_000 ? 250 : 500,
		long_tasks: [],
		browser_errors: [],
		browser_perf_logs: [],
		event_trace: [],
		transport: null,
		error: null
	};
	let browser;
	try {
		await waitForHealth(baseUrl, child);
		const browserExecutable = process.env.GC_UI_BROWSER_EXECUTABLE?.trim();
		const browserChannel =
			process.env.GC_UI_BROWSER_CHANNEL?.trim() ||
			(process.platform === 'win32' ? 'msedge' : undefined);
		browser = await chromium.launch({
			headless: true,
			...(browserExecutable ? { executablePath: browserExecutable } : {}),
			...(!browserExecutable && browserChannel ? { channel: browserChannel } : {})
		});
		report.browser_version = browser.version();
		const page = await browser.newPage({
			viewport: { width: 1600, height: 1000 },
			reducedMotion: 'reduce'
		});
		let captureEventTrace = false;
		if (process.env.GC_CAPTURE_EVENT_TRACE === '1') {
			page.on('websocket', (socket) => {
				socket.on('framereceived', ({ payload }) => {
					if (
						!captureEventTrace ||
						typeof payload !== 'string' ||
						!payload.startsWith('{"kind":"delta"')
					)
						return;
					const message = JSON.parse(payload);
					for (const delta of message.deltas ?? []) {
						for (const event of delta.batch?.events ?? []) {
							if (report.event_trace.length >= 1000) return;
							if (event.kind === 'paramChanged') {
								report.event_trace.push({
									plane: delta.plane,
									time: event.time,
									kind: 'paramChanged',
									param: event.param
								});
							} else if (event.kind === 'graphTransaction') {
								report.event_trace.push({
									plane: delta.plane,
									time: event.time,
									kind: 'graphTransaction',
									ops: event.ops?.map((op) => ({
										kind: op.kind,
										root: op.root,
										inserted: op.nodes?.length
									}))
								});
							}
						}
					}
				});
			});
		}
		page.on('console', (message) => {
			if (message.type() === 'error') report.browser_errors.push(message.text());
			if (
				message.text().startsWith('[ui ws]') ||
				message.text().startsWith('[ui graph]') ||
				message.text().startsWith('[ui-perf] [ui] snapshot') ||
				message.text().startsWith('[ui-perf] [ui] batch events=')
			) {
				if (report.browser_perf_logs.length < 256) report.browser_perf_logs.push(message.text());
			}
		});
		page.on('pageerror', (error) => report.browser_errors.push(error.message));
		await page.addInitScript((profileUi) => {
			if (profileUi) {
				localStorage.setItem('gc_ui_perf', '1');
				localStorage.setItem('gc_perf_profiler', '1');
			}
			globalThis.__gcLongTasks = [];
			new PerformanceObserver((entries) => {
				for (const entry of entries.getEntries()) {
					globalThis.__gcLongTasks.push({ start_ms: entry.startTime, duration_ms: entry.duration });
				}
			}).observe({ type: 'longtask', buffered: true });
		}, process.env.GC_UI_PERF_TRACE === '1');
		await page.goto(`${baseUrl}/?gc_debug_runtime=0`, {
			waitUntil: 'domcontentloaded',
			timeout: 180_000
		});
		await page.locator('.gc-loading-overlay').waitFor({ state: 'hidden', timeout: 180_000 });
		await request(baseUrl, '/api/ui/project-load', { path: fixture, recover: false });
		await page.waitForFunction(
			(target) =>
				Number(document.querySelector('.gc-main')?.getAttribute('data-graph-node-count')) >= target,
			minimumNodes,
			{ timeout: 180_000 }
		);
		const snapshot = await request(baseUrl, '/api/ui/snapshot', { scope: 'wholeGraph' });
		report.base_nodes = snapshot.nodes.length;
		report.base_dom = await page.evaluate(() => ({
			outliner_items: document.querySelectorAll('.outliner-item').length,
			outliner_projected_rows: Number(
				document.querySelector('.outliner-tree')?.getAttribute('data-outliner-row-count')
			),
			canvas_visible_nodes: Array.from(document.querySelectorAll('[data-visible-node-count]')).map(
				(node) => Number(node.getAttribute('data-visible-node-count'))
			)
		}));
		const duplicate = chooseDuplicateSources(snapshot);
		report.duplicated_roots_per_sample = duplicate.nodes.length;
		report.inserted_nodes_per_sample = duplicate.insertedNodes;
		await page.evaluate(() => {
			globalThis.__gcLongTasks = [];
		});
		for (let sample = 0; sample < samples; sample += 1) {
			captureEventTrace = true;
			let cpuProfiler;
			if (sample === 0 && process.env.GC_CAPTURE_CPU_PROFILE === '1') {
				cpuProfiler = await page.context().newCDPSession(page);
				await cpuProfiler.send('Profiler.enable');
				await cpuProfiler.send('Profiler.setSamplingInterval', { interval: 1000 });
				await cpuProfiler.send('Profiler.start');
			}
			const baseCount = await page.evaluate(() =>
				Number(document.querySelector('.gc-main')?.getAttribute('data-graph-node-count'))
			);
			await page.evaluate((base) => {
				const root = document.querySelector('.gc-main');
				if (!root) throw new Error('workbench root disappeared');
				globalThis.__gcPaintMeasurement = new Promise((resolve, reject) => {
					const started = performance.now();
					const timer = setTimeout(() => {
						observer.disconnect();
						reject(new Error('graph projection timed out'));
					}, 180_000);
					const observer = new MutationObserver(() => {
						const count = Number(root.getAttribute('data-graph-node-count'));
						if (count <= base) return;
						const mutationMs = performance.now() - started;
						observer.disconnect();
						clearTimeout(timer);
						requestAnimationFrame(() =>
							requestAnimationFrame(() =>
								resolve({
									latency_ms: performance.now() - started,
									mutation_ms: mutationMs,
									node_count: count
								})
							)
						);
					});
					observer.observe(root, { attributes: true, attributeFilter: ['data-graph-node-count'] });
				});
			}, baseCount);
			const httpStarted = performance.now();
			const acknowledgement = await request(baseUrl, '/api/ui/intent', {
				kind: 'duplicateNodes',
				nodes: duplicate.nodes
			});
			report.http_ack_ms.push(performance.now() - httpStarted);
			if (acknowledgement.success !== true) {
				throw new Error(
					`batch duplicate rejected: ${JSON.stringify(acknowledgement).slice(0, 500)}`
				);
			}
			const painted = await page.evaluate(() => globalThis.__gcPaintMeasurement);
			if (painted.node_count !== baseCount + duplicate.insertedNodes) {
				throw new Error(
					`expected ${baseCount + duplicate.insertedNodes} nodes after batch; saw ${painted.node_count}`
				);
			}
			report.latencies_ms.push(painted.latency_ms);
			report.mutation_ms.push(painted.mutation_ms);
			report.outliner_items_per_sample.push(await page.locator('.outliner-item').count());
			report.outliner_projected_rows_per_sample.push(
				Number(await page.locator('.outliner-tree').getAttribute('data-outliner-row-count'))
			);
			if (cpuProfiler) {
				const { profile } = await cpuProfiler.send('Profiler.stop');
				await writeFile(join(outputDir, 'browser-action.cpuprofile'), JSON.stringify(profile));
				await cpuProfiler.detach();
			}
		}
		if (process.env.GC_VERIFY_OUTLINER_SCROLL === '1') {
			const before = await page.evaluate(() => {
				const scroller = document.querySelector('.outliner-content');
				if (!(scroller instanceof HTMLElement)) throw new Error('outliner scroller is missing');
				const first = scroller.querySelector('.outliner-item-content[data-node-id]');
				const firstId = first?.getAttribute('data-node-id');
				if (!firstId || scroller.scrollHeight <= scroller.clientHeight) {
					throw new Error('outliner has no scrollable visible rows');
				}
				return {
					firstId,
					scrollHeight: scroller.scrollHeight,
					clientHeight: scroller.clientHeight
				};
			});
			await page.evaluate(() => {
				const scroller = document.querySelector('.outliner-content');
				if (!(scroller instanceof HTMLElement)) throw new Error('outliner scroller is missing');
				scroller.scrollTop = scroller.scrollHeight / 2;
			});
			await page.waitForFunction(
				(firstId) =>
					document
						.querySelector('.outliner-content .outliner-item-content[data-node-id]')
						?.getAttribute('data-node-id') !== firstId,
				before.firstId,
				{ timeout: 30_000 }
			);
			const after = await page.evaluate(() => {
				const scroller = document.querySelector('.outliner-content');
				if (!(scroller instanceof HTMLElement)) throw new Error('outliner scroller is missing');
				return {
					firstId: scroller
						.querySelector('.outliner-item-content[data-node-id]')
						?.getAttribute('data-node-id'),
					items: scroller.querySelectorAll('.outliner-item').length,
					scrollTop: scroller.scrollTop
				};
			});
			if (after.items > 200 || after.scrollTop <= 0) {
				throw new Error(`outliner scroll window is unbounded: ${JSON.stringify(after)}`);
			}
			report.outliner_scroll_probe = { before, after, status: 'PASS' };
		}
		if (process.env.GC_VERIFY_OUTLINER_META === '1') {
			await page.evaluate(() => {
				const scroller = document.querySelector('.outliner-content');
				if (!(scroller instanceof HTMLElement)) throw new Error('outliner scroller is missing');
				scroller.scrollTop = 0;
			});
			if (report.outliner_scroll_probe) {
				await page.waitForFunction(
					(firstId) =>
						document
							.querySelector('.outliner-content .outliner-item-content[data-node-id]')
							?.getAttribute('data-node-id') === firstId,
					report.outliner_scroll_probe.before.firstId,
					{ timeout: 30_000 }
				);
			}
			const mountedIds = await page.evaluate(() =>
				Array.from(document.querySelectorAll('.outliner-item-content[data-node-id]')).map((row) =>
					Number(row.getAttribute('data-node-id'))
				)
			);
			const nodesById = new Map(snapshot.nodes.map((node) => [node.node_id, node]));
			const source = mountedIds.find(
				(nodeId) => nodesById.get(nodeId)?.meta?.user_permissions?.can_edit_name === true
			);
			if (source === undefined) throw new Error('no editable outliner row is mounted');
			const label = '__outliner_projection_probe__';
			const acknowledgement = await request(baseUrl, '/api/ui/intent', {
				kind: 'patchMeta',
				node: source,
				patch: { label }
			});
			if (acknowledgement.success !== true) {
				throw new Error(`outliner metadata probe rejected: ${JSON.stringify(acknowledgement)}`);
			}
			await page.waitForFunction(
				({ source, label }) =>
					document
						.querySelector(`.outliner-item-content[data-node-id="${source}"] .outliner-item-label`)
						?.textContent?.trim() === label,
				{ source, label },
				{ timeout: 30_000 }
			);
			report.outliner_meta_probe = { source, label, status: 'PASS' };
		}
		if (process.env.GC_CAPTURE_OUTLINER_SCREENSHOT === '1') {
			await page.locator('.outliner-content').screenshot({
				path: join(outputDir, 'outliner-viewport.png')
			});
		}
		report.long_tasks = await page.evaluate(() => globalThis.__gcLongTasks);
		report.p50_ms = percentile(report.latencies_ms, 0.5);
		report.p95_ms = percentile(report.latencies_ms, 0.95);
		report.p99_ms = percentile(report.latencies_ms, 0.99);
		report.max_ms = Math.max(...report.latencies_ms);
		if (report.browser_errors.length > 0) throw new Error('browser reported errors');
		if (report.p95_ms > report.p95_target_ms)
			throw new Error('p95 exceeded the provisional workbench budget');
		report.status = 'PASS';
	} catch (error) {
		report.error = String(error?.stack ?? error);
	} finally {
		await browser?.close().catch(() => {});
		if (child.exitCode === null) {
			child.kill();
			await Promise.race([new Promise((done) => child.once('exit', done)), pause(5_000)]);
		}
		report.transport = {
			slow_client_disconnects: (serverLog.match(/disconnecting slow client/g) ?? []).length,
			overflow_resyncs: (serverLog.match(/pausing overloaded subscription/g) ?? []).length,
			ws_snapshots: (serverLog.match(/\[ui-ws\] snapshot request_id=/g) ?? []).length
		};
		await writeFile(join(outputDir, 'headless-server.log'), serverLog);
		await writeFile(
			join(outputDir, 'workbench-paint-report.json'),
			`${JSON.stringify(report, null, 2)}\n`
		);
	}
	return report;
}

const report = await run();
console.log(JSON.stringify({ status: report.status, p95_ms: report.p95_ms, error: report.error }));
if (report.status !== 'PASS') process.exitCode = 1;

File: tools/qualification/live_workbench_paint.py
"""Measure authored graph batch edits through the bundled headless product and browser."""

from __future__ import annotations

import argparse
import json
import math
import os
import platform
import shutil
import subprocess
import sys
import tomllib
from collections.abc import Sequence
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

if __package__:
    from .formula_scale import cargo_environment
    from .graph_fixture import DEFAULT_SOURCE, write_fixture
    from .runtime_scale import command_output, sha256_bytes, utc_now, working_tree_sha
else:
    from formula_scale import cargo_environment
    from graph_fixture import DEFAULT_SOURCE, write_fixture
    from runtime_scale import command_output, sha256_bytes, utc_now, working_tree_sha


CONTRACT = "chataigne-live-workbench-paint-v1"
TARGETS = (10_000, 100_000)
BUILD_COMMAND = (
    "cargo", "build", "--locked", "-p", "Chataigne2", "--bin", "Chataigne2",
    "--target-dir", "target/t16-app-default",
)
PROBE_SCRIPT = Path("tools/qualification/live_workbench_paint.mjs")


def percentile(values: list[float], quantile: float) -> float:
    ordered = sorted(values)
    return ordered[min(len(ordered) - 1, math.ceil(len(ordered) * quantile) - 1)]


def parse_probe_report(
    report: dict[str, Any], target: int, samples: int, binary_sha: str, fixture_sha: str,
) -> str:
    if report.get("contract") != CONTRACT or report.get("minimum_nodes") != target:
        raise ValueError("live workbench report contract or fixture target differs")
    if report.get("samples") != samples or report.get("binary_sha256") != binary_sha \
            or report.get("fixture_sha256") != fixture_sha:
        raise ValueError("live workbench report fingerprint or sample count differs")
    if type(report.get("base_nodes")) is not int or report["base_nodes"] < target:
        raise ValueError("live workbench never reached the authored node target")
    if type(report.get("duplicated_roots_per_sample")) is not int \
            or report["duplicated_roots_per_sample"] < 1 \
            or type(report.get("inserted_nodes_per_sample")) is not int \
            or report["inserted_nodes_per_sample"] < 600:
        raise ValueError("live workbench did not perform a 600-node batch")
    for field in ("latencies_ms", "http_ack_ms", "mutation_ms"):
        values = report.get(field)
        if not isinstance(values, list) or len(values) != samples \
                or any(type(value) not in (float, int) or not math.isfinite(value) or value < 0
                       for value in values):
            raise ValueError(f"live workbench has incomplete {field}")
    latencies = report["latencies_ms"]
    for field, quantile in (("p50_ms", 0.5), ("p95_ms", 0.95), ("p99_ms", 0.99)):
        if report.get(field) != percentile(latencies, quantile):
            raise ValueError(f"live workbench {field} does not match raw samples")
    if report.get("max_ms") != max(latencies):
        raise ValueError("live workbench max does not match raw samples")
    target_ms = 250 if target == 10_000 else 500
    if report.get("p95_target_ms") != target_ms:
        raise ValueError("live workbench silently changed the provisional p95 budget")
    if not isinstance(report.get("long_tasks"), list) or not isinstance(report.get("browser_errors"), list) \
            or not isinstance(report.get("browser_perf_logs"), list):
        raise ValueError("live workbench browser diagnostics are incomplete")
    if report["browser_errors"]:
        raise ValueError("live workbench recorded browser errors")
    transport = report.get("transport")
    if not isinstance(transport, dict) or set(transport) != {
        "slow_client_disconnects", "overflow_resyncs", "ws_snapshots",
    } or any(type(value) is not int or value < 0 for value in transport.values()):
        raise ValueError("live workbench transport recovery counters are incomplete")
    if transport["slow_client_disconnects"] != 0:
        raise ValueError("live workbench still disconnected a slow browser client")
    budget_passed = report["p95_ms"] <= target_ms
    if budget_passed and report.get("status") == "PASS" and report.get("error") is None:
        return "PASS"
    if not budget_passed and report.get("status") == "FAIL" \
            and "p95 exceeded the provisional workbench budget" in str(report.get("error")):
        return "BUDGET_FAIL"
    raise ValueError("live workbench report status does not match its measurements")


def resolve_output_dir(root: Path, value: Path | None) -> Path:
    target_root = (root / "target").resolve()
    output = (
        value.resolve() if value is not None and value.is_absolute()
        else (root / value).resolve() if value is not None
        else target_root / "qualification" / "live-workbench-paint"
        / datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    )
    if not output.is_relative_to(target_root) or output == target_root:
        raise ValueError("live workbench output must be a child of the workspace target directory")
    if output.exists() and any(output.iterdir()):
        raise ValueError("live workbench output directory already contains artifacts")
    return output


def build_report(root: Path, output_dir: Path, samples: int) -> dict[str, Any]:
    started_at = utc_now()
    tested_tree_sha = working_tree_sha(root)
    commit_sha = command_output(root, ("git", "rev-parse", "HEAD"))
    node = shutil.which("node")
    manifest = tomllib.loads((root / "apps/chataigne/Cargo.toml").read_text(encoding="utf-8"))
    source_sha = sha256_bytes((root / DEFAULT_SOURCE).read_bytes())
    environment = cargo_environment(root)
    environment.pop("GC_SKIP_UI_BUILD", None)
    output_dir.mkdir(parents=True, exist_ok=True)

    build = subprocess.run(
        BUILD_COMMAND, cwd=root, env=environment, capture_output=True,
        check=False, text=True, encoding="utf-8",
    )
    build_log = output_dir / "cargo-build.log"
    build_log.write_text(build.stdout + build.stderr, encoding="utf-8")
    binary = root / "target/t16-app-default/debug" / ("Chataigne2.exe" if os.name == "nt" else "Chataigne2")
    binary_sha = sha256_bytes(binary.read_bytes()) if build.returncode == 0 and binary.is_file() else None

    scenarios = []
    for target in TARGETS:
        fixture = output_dir / f"authored-{target}.noisette"
        metadata = write_fixture(root / DEFAULT_SOURCE, fixture, minimum_live_nodes=target)
        fixture_sha = sha256_bytes(fixture.read_bytes())
        probe_dir = output_dir / f"workbench-{target}"
        command = (node or "node", str(root / PROBE_SCRIPT), str(binary), str(fixture), str(target),
                   str(probe_dir), str(samples))
        process = None
        if binary_sha is not None and node is not None:
            process = subprocess.run(
                command, cwd=root, env=environment, capture_output=True, check=False,
                text=True, encoding="utf-8",
            )
        output = "" if process is None else process.stdout + process.stderr
        log = output_dir / f"workbench-{target}.log"
        log.write_text(output, encoding="utf-8")
        probe_report_path = probe_dir / "workbench-paint-report.json"
        measured = None
        parse_error = None
        status = "MISSING"
        if probe_report_path.is_file() and binary_sha is not None:
            try:
                measured = json.loads(probe_report_path.read_text(encoding="utf-8"))
                status = parse_probe_report(measured, target, samples, binary_sha, fixture_sha)
            except (ValueError, json.JSONDecodeError) as error:
                parse_error = str(error)
        else:
            parse_error = "probe report missing or product build failed"
        if process is not None and ((status == "PASS" and process.returncode != 0)
                                    or (status == "BUDGET_FAIL" and process.returncode == 0)):
            status = "MISSING"
            parse_error = "probe exit code disagrees with validated report"
        scenarios.append({
            "target": target, "status": status, "command": list(command),
            "exit_code": None if process is None else process.returncode,
            "fixture": {**metadata, "sha256": fixture_sha},
            "log": {"path": log.relative_to(root).as_posix(), "sha256": sha256_bytes(output.encode())},
            "probe_report": probe_report_path.relative_to(root).as_posix() if probe_report_path.is_file() else None,
            "measured_result": measured, "parse_error": parse_error,
        })
        print(f"live workbench {target}: {status} ({log})", flush=True)

    if working_tree_sha(root) != tested_tree_sha:
        raise ValueError("source tree changed during live workbench qualification")
    return {
        "schema_version": 1,
        "evidence_id": "product.live-workbench-paint.local",
        "status": "PASS" if build.returncode == 0 and all(row["status"] == "PASS" for row in scenarios) else "FAIL",
        "product_qualification": "OPEN",
        "commit_sha": commit_sha, "tested_tree_sha": tested_tree_sha,
        "artifact": {"path": binary.relative_to(root).as_posix(), "sha256": binary_sha},
        "build": {"command": list(BUILD_COMMAND), "exit_code": build.returncode,
                  "log": build_log.relative_to(root).as_posix()},
        "features": {"default": manifest["features"]["default"], "ui_assets_skipped": False},
        "toolchain_fingerprint": {
            "rustc": command_output(root, ("rustc", "-Vv")),
            "cargo": command_output(root, ("cargo", "-V")),
            "node": command_output(root, ("node", "--version")) if node else None,
            "os": platform.platform(),
        },
        "source_fixture": {"path": DEFAULT_SOURCE.as_posix(), "sha256": source_sha},
        "started_at": started_at, "finished_at": utc_now(),
        "scenarios": scenarios,
        "scope": "one real HTTP batch duplicate of at least 600 nodes per sample; bundled-UI headless app, "
                 "workbench WebSocket projection, DOM mutation, and two browser frame callbacks",
        "not_covered": ["direct pixel-difference proof", "native webview and physical devices"],
    }


def main(arguments: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", type=Path, default=Path.cwd())
    parser.add_argument("--output-dir", type=Path)
    parser.add_argument("--samples", type=int, default=20)
    args = parser.parse_args(arguments)
    if args.samples < 1:
        parser.error("--samples must be positive")
    root = args.root.resolve()
    output_dir = resolve_output_dir(root, args.output_dir)
    report = build_report(root, output_dir, args.samples)
    report_path = output_dir / "live-workbench-paint-report.json"
    report_path.write_text(json.dumps(report, indent=2) + "\n", encoding="utf-8")
    print(f"Live workbench report: {report_path.relative_to(root)}", flush=True)
    return 0 if report["status"] == "PASS" else 1


if __name__ == "__main__":
    sys.exit(main())
